src/util/Serializer/CompressedSerializer.h

```diff
@@ -0,0 +1,282 @@
+// Copyright 2021 - 2025 The QLever Authors, in particular:
+//
+// 2021 Robin Textor-Falconi <textorr@cs.uni-freiburg.de>, UFR
+// 2025 Johannes Kalmbach <kalmbach@cs.uni-freiburg.de>, UFR
+
+// UFR = University of Freiburg, Chair of Algorithms and Data Structures
+
+// You may not use this file except in compliance with the Apache 2.0 License,
+// which can be found in the `LICENSE` file at the root of the QLever project.
+
+#ifndef QLEVER_SRC_UTIL_SERIALIZER_COMPRESSEDSERIALIZER_H
+#define QLEVER_SRC_UTIL_SERIALIZER_COMPRESSEDSERIALIZER_H
+
+#include <functional>
+#include <optional>
+#include <vector>
+
+#include "backports/span.h"
+#include "util/CompressionUsingZstd/ZstdWrapper.h"
+#include "util/ExceptionHandling.h"
+#include "util/MemorySize/MemorySize.h"
+#include "util/Serializer/SerializeVector.h"
+#include "util/Serializer/Serializer.h"
+#include "util/TypeTraits.h"
+#include "util/UninitializedAllocator.h"
+
+namespace ad_utility::serialization {
+
+// A `vector<char>` where a `resize` doesn't zero-initialize the new bytes
+// (for efficiency reasons).
+using UninitializedBuffer =
+    std::vector<char, ad_utility::default_init_allocator<char>>;
+
+/**
+ * A write serializer that compresses data in blocks before writing to an
+ * underlying serializer.
+ *
+ * Data is buffered until the buffer exceeds the block size, then the full
+ * block is compressed and serialized as a vector to the underlying serializer.
+ * The last (possibly incomplete) block is written on destruction or when
+ * close() is called.
+ *
+ * @tparam UnderlyingSerializer The underlying serializer type (must be a
+ * WriteSerializer)
+ * @tparam CompressionFunction A callable that takes a span of chars and returns
+ * a vector of chars
+ */
+CPP_template(typename UnderlyingSerializer, typename CompressionFunction)(
+    requires WriteSerializer<UnderlyingSerializer> CPP_and ql::concepts::
+        invocable<CompressionFunction, ql::span<const char>,
+                  UninitializedBuffer&>) class CompressedWriteSerializer {
```

Comment:
These are really hard to read. How hard is it to patch or configure clang-format to format this properly?

Reply:
Thank you for the reminder, we would have to update + configure clang-format.